Swift option added for pfioh, test suite added and readme added

awalkaradi95moc opened the pull request:
@danmcp This is the new pull request having changes suggested in previous one.

danmcp commented on bin/pfioh:
```diff
@@ -21,11 +21,16 @@ import  pudb
 try:
-    import  pfioh
+    import pfioh
+    import mount_dir
```
Just move the import of mount_dir and swift_store out of the try catch entirely.

danmcp commented on swift-readme.md:
```diff
@@ -0,0 +1,30 @@
+##Swift Object Store
```
Maybe put this in the openshift readme?

https://github.com/FNNDSC/pfioh/blob/master/openshift/README.rst

danmcp commented:
@rudolphpienaar Can you take a look at this and let us know what you think?

danmcp commented on pfioh/swift_store.py:
```diff
@@ -0,0 +1,266 @@
+            return d_ret
+
+        #Delete temporary empty directory created by Swift
+        self._deleteEmptyDirectory(key)
```
Should this be in a finally block?

danmcp commented on pfioh/swift_store.py:
```diff
@@ -0,0 +1,266 @@
+        with open('ziparchive.zip','rb') as f:
+            zippedFileContent = f.read()
+
+        os.remove('ziparchive.zip')
```
Should this be in a finally block?

danmcp commented on pfioh/swift_store.py:
```diff
@@ -0,0 +1,266 @@
+        config= configparser.ConfigParser()
+        f = open(configPath, 'r')
+        config.readfp(f)
+        f.close()
```
Should be in a finally block

danmcp commented on pfioh/swift_store.py:
```diff
@@ -0,0 +1,266 @@
+
+        try:
+            self.swiftConnection.put_container(key)
+            self.qprint('Container added sucessfully')
```
successfully

danmcp replied:
Also maybe change the comment to swift container added successfully to make it clear what kind of container you are talking about.

danmcp commented on pfioh/swift_store.py:
```diff
@@ -0,0 +1,266 @@
+        except Exception as err:
+            self.qprint(err)
+            d_ret['status'] = False
+            d_ret['msg']    = 'Retriving File/Directory from Swift failed'
```
Retrieving

danmcp commented on swift-readme.md:
```diff
@@ -0,0 +1,30 @@
+2) Now run the following command to create a secret
+`oc create secret generic swift-credentials --from-file=<path-to-file>/swift-credentials.cfg`
+
+3) Now attach this secret to the pod, by clicking *add config files to <dc-name>* 
```
Is this for the UI?  Rather than do this, add the secret reference in the template:

https://github.com/FNNDSC/pfioh/blob/master/openshift/pfioh-openshift-template.json

awalkaradi95moc replied:
Yes this is for UI. If user is not using the provided template which has the secret, he has to follow these steps. Also, I'm adding the secret into the template.

danmcp commented on pfioh/mount_dir.py:
```diff
@@ -0,0 +1,176 @@
+        d_ret = self.readData(str_fileToProcess, d_ret)
+
+        #Cleanup by deleting temporary files
+        if b_cleanup:
```
this should really be in a finally block as well.

danmcp commented on pfioh/mount_dir.py:
```diff
@@ -0,0 +1,176 @@
+
+    def readData(self, str_fileToProcess, d_ret):
+        """
+        Reads the data from the file and writes it for transfering over the network
```
transferring

danmcp commented on pfioh/mount_dir.py:
```diff
@@ -0,0 +1,176 @@
+            d_ret['status'] = d_fio['status']
+            d_ret['msg']    = d_fio['msg']
+
+            os.remove(fileName)
```
should be in a finally block

danmcp commented on pfioh/swift_store.py:
```diff
@@ -0,0 +1,266 @@
+
+    def _initiateSwiftConnection(self, **kwargs):
+        """
+        Initiates a Swift connection and returns a Swift Connection Object
```
lowercase connection object

danmcp commented on openshift/README.rst:
```diff
@@ -4,4 +4,34 @@ Setup:
+2) Now run the following command to create a secret
+`oc create secret generic swift-credentials --from-file=<path-to-file>/swift-credentials.cfg`
+
+3) Now attach this secret to the pod, by clicking *add config files to <dc-name>* 
```
I would just take these out and have the secret reference in the template.

awalkaradi95moc replied:
But the file format is not standard, rather we have defined it. Hence somewhere we need to specify that how are we expecting the swift-credentials.cfg file to be. This is not the case for .kube/config since the config file here is created by Openshift on oc login and has a standard format.

danmcp replied:
I am referring to step 3 and 4.  Are you as well?

awalkaradi95moc replied:
No, I thought you were referring to the complete swift readme. Yes, it makes sense deleting steps 3 and 4.

danmcp commented on pfioh/mount_dir.py:
```diff
@@ -0,0 +1,179 @@
+                d_ret['unzip']  = d_fio
+                d_ret['status'] = d_fio['status']
+                d_ret['msg']    = d_fio['msg']
+            finally:
```
I think this is at the wrong place.  It needs to correspond to the open above.

awalkaradi95moc replied:
I didn't understand this. If I move it, to correspond with open, it will delete the zip file before unpacking.

danmcp replied:
Move the try up to the open.  And put the finally just above the if b_zip.

danmcp commented on pfioh/swift_store.py:
```diff
@@ -0,0 +1,269 @@
+        for k,v in kwargs:
+            if k == 'configPath': configPath= v
+
+        configPath= '/etc/swift/swift-credentials.cfg'
```
space before and after the = for each one of these.

rudolphpienaar commented:
In future, please make sure that PR requests actually run w/o error. There were several dependency issues I had to resolve as well as syntax errors (or at least on my dev setup).

It would be great if you can make sure that whatever changes you make don't break the docker image (just do a local docker image build and see if the container actually runs).

The correct (albeit convoluted) way to import local python modules into pfioh is:

```
# The following convolutions are necessary to accommodate both
# development and deployed instances of the 'pfioh.py' module
# and related dependencies.
sys.path.insert(1, os.path.join(os.path.dirname(__file__), '../pfioh'))
try:
    import pfioh
    import mount_dir
    import swift_store
except:
    sys.path.insert(1, os.path.join(get_python_lib(True, True), 'site-packages'))
    sys.path.insert(1, os.path.join(get_python_lib(True, True), 'site-packages/pfioh'))
    import pfioh
    import mount_dir
    import swift_store
```

Also, I had to manually do 

```
pip3 install keystoneauth1
pip3 install python-swiftclient
```

with corresponding changes to the Dockerfile, too:

```
  && pip3 install pudb                                                \
  && pip3 install keystoneauth1                                       \
  && pip3 install python-swiftclient
```

awalkaradi95moc commented:
@rudolphpienaar I had ran the complete code locally and it had worked. Strangely I am also getting module not found error in pfioh/bin while running on Openshift, while it did work locally.  So I had to change it to 'from pfioh import mount_dir'. I believe since I was inserting pfioh/pfioh in python path, it should have directly imported mount_dir and swift_store, not sure why that didn't work on Openshift. 

Also, the other modules like pudb, keystoneauth1 and python-swiftclient I had added in setup.py which installed those modules for me and docker build was successful. Now when I tried to figure out, I understood what was happening. I had changed the default pfioh version to /tmp/work so as to install the latest pfioh which I have built
pip3 install --prefix /usr pfioh==1.4.5 was changed to pip3 install /tmp/work 
and this used to run the setup.py file

I should have thought about this. I will be extra careful next time for all such errors. I am sorry for the trouble.

rudolphpienaar commented:
Rejected this PR.

Integration tests failed on changes to pfioh. On deeper examination, noticed some what appeared to be syntax issues in mount_dir. Odd that tests on @awalkaradi95moc setup didn't bomb out like they did on mine.

Also, logic was rewritten from pfioh when "moved" to mount_dir. Certain logic flow issues were noted. Furthermore, cleanup directives in the JSON 'transport' paragraph were ignored and other changes noted.

Right now, replaced pfioh/pfioh.py with the latest working version (from repo with internal ver 1.4.6) and suggested to @awalkaradi95moc  to also adopt some Hungarian naming convention style in variable naming, etc.
